pkg/cloudproviders/k8s/k8s.go

```diff
@@ -0,0 +1,348 @@
+	var restConfig *rest.Config
+	var err error
+	if kubeconfigPath != "" {
+		configOverrides := &clientcmd.ConfigOverrides{}
+		if kubeconfigContext != "" {
+			configOverrides.CurrentContext = kubeconfigContext
+		}
+
+		restConfig, err = clientcmd.NewNonInteractiveDeferredLoadingClientConfig(
+			&clientcmd.ClientConfigLoadingRules{ExplicitPath: kubeconfigPath},
+			configOverrides).ClientConfig()
+	} else {
+		restConfig, err = rest.InClusterConfig()
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	restConfig.RateLimiter = flowcontrol.NewTokenBucketRateLimiter(float32(apiQPS), int(apiQPS*apiQPSBurstFactor))
+	restConfig.UserAgent = userAgent
+
+	clientset, err := kubernetes.NewForConfig(restConfig)
+	if err != nil {
+		return nil, err
+	}
+	return clientset, nil
+}
+
+func podByIpIndexFunc(obj interface{}) ([]string, error) {
+	pod := obj.(*core_v1.Pod)
+	ip := pod.Status.PodIP
+	// Pod must have an IP, be running, and also not have the same IP as the host
+	// If a pod is a HostNetwork pod then there could be multiple with the same IP sending stats, which breaks this
+	// abstraction
+	if ip == "" ||
+		podIsFinishedRunning(pod) ||
+		podIsHostNetwork(pod) {
+		// Do not index irrelevant Pods
+		return nil, nil
+	}
+	return []string{ip}, nil
+}
+
+func podIsHostNetwork(pod *core_v1.Pod) bool {
+	return pod.Spec.HostNetwork || pod.Status.PodIP == pod.Status.HostIP
+}
+
+func podIsFinishedRunning(pod *core_v1.Pod) bool {
+	return pod.Status.Phase == core_v1.PodSucceeded || pod.Status.Phase == core_v1.PodFailed || pod.DeletionTimestamp != nil
+}
+
+func setViperDefaults(v *viper.Viper, version string) {
+	v.SetDefault(ParamAPIQPS, DefaultAPIQPS)
+	v.SetDefault(ParamAPIQPSBurstFactor, DefaultAPIQPSBurstFactor)
+	v.SetDefault(ParamAnnotationTagRegex, DefaultAnnotationTagRegex)
+	v.SetDefault(ParamKubeconfigContext, DefaultKubeconfigContext)
+	v.SetDefault(ParamKubeconfigPath, DefaultKubeconfigPath)
+	v.SetDefault(ParamLabelTagRegex, DefaultLabelTagRegex)
+	// This is intended to be taken in primarily as an environment variable when running inside k8s, as that is the
+	// k8s standard way of providing variable information to pods via the downwards API.
+	// See: https://kubernetes.io/docs/tasks/inject-data-application/environment-variable-expose-pod-information/
+	v.SetDefault(ParamNodeName, DefaultNodeName)
+	v.SetDefault(ParamResyncPeriod, DefaultResyncPeriod)
+	v.SetDefault(ParamUserAgent, DefaultUserAgent+"/"+version)
+	v.SetDefault(ParamWatchCluster, DefaultWatchCluster)
+}
+
+// getTagNameFromRegex gets a tag name from the regex. This is either the entire input string, or a subset matching
+// the "tag" capture group if it exists and matched. Returns "" if the regex matched nothing.
+func getTagNameFromRegex(re *regexp.Regexp, s string) string {
+	match := re.FindStringSubmatch(s)
+
+	var matches = make(map[string]string, len(match))
```

Comment:
We don't need this map do we? Can just do `switch` inside of the loop and look for interesting group names. WDYT?

Reply:
If we're concerned about GC then yeah I can change it for that reason. It does complicate the code a little though since there's two slices in question and we'd have to loop through them in reverse to find the subexp match before the entire regex.